File: StringCalculator/StringCalculator.fs
﻿module StringCalculator



let ExtractDelimiters (delimiters : string array, numbersString : string) = 
    if numbersString.StartsWith("//") then
        let inputs = numbersString.Split([|'\n'|],2)
        let customDelimiters = inputs.[0].Split([|"["; "]"; "/"|],System.StringSplitOptions.RemoveEmptyEntries)
        (Array.append delimiters customDelimiters,inputs.[1])
    else
        (delimiters,numbersString)

exception InvalidInput of string
let CheckInvaildInput (numbersString : string array) =
    let emptyItems = Array.filter(fun x -> System.String.IsNullOrEmpty(x)) numbersString;
    if emptyItems.Length > 0 then
        raise(InvalidInput "Invaild input format")
    numbersString

let ExtractNumbers (delimiters : string array, numbersString : string) =
    if System.String.IsNullOrEmpty(numbersString) then
        [||]
    else
    numbersString.Split(delimiters, System.StringSplitOptions.None)
    |> CheckInvaildInput
    |> Array.map int

exception NegativesNoAllowed of int[]
let NegativeCheck (numbers : int array) =
    let negativeNumbers = Array.filter(fun n -> n < 0) numbers 
    if negativeNumbers.Length > 0 then
        raise(NegativesNoAllowed negativeNumbers)
    numbers

let FliterNumbersBiggerThan1000 (numbers : int array) =
     Array.filter(fun n -> n <= 1000) numbers

let Add (numbers : string) = 
    let  delimiters = [|",";"\n"|]
    if System.String.IsNullOrEmpty(numbers) then
        0
    else
    ExtractDelimiters(delimiters,numbers)
    |> ExtractNumbers
    |> NegativeCheck
    |> FliterNumbersBiggerThan1000
    |> Array.sum
       

File: StringCalculatorCS/StringCalculator.cs
namespace StringCalculatorCS
{
    internal static class StringCalculator
    {
        public static List<string> Delimiters { get; } = [",", "\n"];
        public static int Add(String numbers)
        {
            if (String.IsNullOrEmpty(numbers)) return 0;
            return FliterNumbersBiggerThan1000(NegativeCheck(ExtractNumbers(ExtractDelimiters(numbers)))).Sum();

        }
        private static string ExtractDelimiters(String numbers)
        {
            if (!numbers.StartsWith("//")) return numbers;
            var inputs = numbers.Split('\n', 2);
            var customDelimiters = inputs[0].Split(new char[] { '[', ']', '/' }, StringSplitOptions.RemoveEmptyEntries);
            Delimiters.AddRange(customDelimiters);
            return inputs[1];
        }

        private static int[] ExtractNumbers(string numbers)
        {
            if (String.IsNullOrEmpty(numbers)) return [];
            return Array.ConvertAll(
                 CheckInvaildInput(
                     numbers.Split(Delimiters.ToArray(), StringSplitOptions.None)
                 ),
                 int.Parse
             );
        }

        private static string[] CheckInvaildInput(string[] numbers)
        {
            var emptyItems = numbers.Where(x => x == "" || x == " ").ToArray();
            if (emptyItems.Length > 0)
            {
                throw new InvalidInputException("Invaild input format");
            }
            return numbers;
        }

        private static int[] NegativeCheck(int[] numbers)
        {
            var negativeNumbers = numbers.Where((x) => x < 0).ToArray();
            if (negativeNumbers.Length > 0)
            {
                throw new NegativesNoAllowedException(negativeNumbers);
            }
            return numbers;
        }

        private static int[] FliterNumbersBiggerThan1000(int[] numbers)
        {
            return numbers.Where((x) => x <= 1000).ToArray();
        }
    }
}
